rotkehlchen/externalapis/defillama.py

```diff
@@ -0,0 +1,290 @@
+    ) -> bool:
+        if seconds is None:
+            return False
+
+        return ts_now() - self.last_rate_limit <= seconds
+
+    def query_historical_price(
+            self,
+            from_asset: Asset,
+            to_asset: Asset,
+            timestamp: Timestamp,
+    ) -> Price:
+        """
+        May raise:
+        - PriceQueryUnsupportedAsset if either from_asset or to_asset are not supported
+        - NoPriceForGivenTimestamp if no price could be found
+        """
+        try:
+            from_asset = from_asset.resolve_to_asset_with_oracles()
+            to_asset = to_asset.resolve_to_asset_with_oracles()
+        except UnknownAsset as e:
+            raise PriceQueryUnsupportedAsset(e.identifier) from e
+
+        # check DB cache
+        price_cache_entry = GlobalDBHandler().get_historical_price(
+            from_asset=from_asset,
+            to_asset=to_asset,
+            timestamp=timestamp,
+            max_seconds_distance=DAY_IN_SECONDS,
+            source=HistoricalPriceOracle.DEFILLAMA,
+        )
+        if price_cache_entry:
+            return price_cache_entry.price
+
+        try:
+            coin_id = self._get_asset_id(from_asset)
+        except UnsupportedAsset as e:
+            log.warning(
+                f'Tried to query historical price using DefiLlama from {from_asset} to '
+                f'{to_asset} at {timestamp} but {from_asset} is not an EVM token and is not '
+                f'suppoorted by coingecko',
+            )
+            raise NoPriceForGivenTimestamp(
+                from_asset=from_asset,
+                to_asset=to_asset,
+                time=timestamp,
+                rate_limited=False,
+            ) from e
+
+        # no cache, query coingecko for daily price
+        date = timestamp_to_date(timestamp, formatstr='%d-%m-%Y')
+        result = self._query(
+            module='prices',
+            subpath=f'historical/{timestamp}/{coin_id}',
+        )
+
+        if 'coins' not in result or len(result['coins']) == 0:
+            log.warning(
+                f'Queried DefiLlama current price from {from_asset.identifier} '
+                f'to {to_asset.identifier}. But coins ins not available in the result {result}',
+            )
+            raise NoPriceForGivenTimestamp(
+                from_asset=from_asset,
+                to_asset=to_asset,
+                time=timestamp,
+                rate_limited=False,
+            )
+
+        coin_result_raw = result['coins'][coin_id]
+        try:
+            if (
+                'confidence' in coin_result_raw and
+                FVal(coin_result_raw['confidence']) < FVal('0.20')
+            ):
+                # DefiLlama provides a confidence value ranking how good they measure the price
+                # is. When their confidence in the price is lower than 20% ignore it. Probably a
+                # spam token
+                return Price(ZERO)
```

Comment:
I start to see lots of similar behavior here, like the confidence calculation along with return etc.

If it makes sense can you abstract what is possible to abstract?
```suggestion
                # DefiLlama provides a confidence value ranking how good their confidence in the price calculation
                # is. When their confidence in the price is lower than 20% ignore it. Probably a
                # spam token
                return Price(ZERO)
```